### PR TITLE
RFC-8: Add TCP protocol

### PR DESCRIPTION
**Pull request by bokket**
Signed-off-by: bokket <3100563328@qq.com>

### CODE REVIEW

File: docs/rfcs/8-add-tcp-pair.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+- Author: bokket  [bokkett@gmail.com](mailto:bokkett@gmail.com)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]:
```suggestion
- Author: bokket <bokkett@gmail.com>
```

File: docs/rfcs/8-add-tcp-pair.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+- Start Date: 2021-07-17
+- RFC PR: [beyondstorage/go-endpoint#8](https://github.com/beyondstorage/go-endpoint/pull/8)
+- Tracking Issue: [beyondstorage/go-endpoint/issues/7](https://github.com/beyondstorage/go-endpoint/issues/7)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]: I think #7 is not a tracking issue, we need to create a new one after this RFC got approved.

File: docs/rfcs/8-add-tcp-pair.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+# RFC-8: Add Tcp pair
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]: There is no `pair` idea in the endpoint, how about renaming it into `Add TCP protocol`?

File: docs/rfcs/8-add-tcp-pair.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+hdfs usually use the `New(address string)` method to access a namenode node, the user will be the user running the code. If the address is an empty string, it will try to get the NameNode address from the Hadoop configuration file.
+
+## Proposal
+
+I suggest adding a pair to allow the user to specify the address.
+
+- The `type` of `tcp` should be `String` and is a `const`
+- The `format` of `ProtocolTcp` should follow  [go-endpoint](https://github.com/beyondstorage/go-endpoint/blob/master/README.md)
+- The `value` of `endpoint` should be parsed into `address include host and port`
+- Now we don't have a pair operation on the `hdfs address` or tcp-like operation
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]: Please describe the format of endpoint directly, like `tcp:<host>:<ip>`.

File: docs/rfcs/8-add-tcp-pair.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+## Rationale
+
+### Why not use hadoop configuration?
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]: It's not related to `go-endpoint` itself, we can remove this section.

File: docs/rfcs/8-add-tcp-pair.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+However, there is no guarantee that the path is wrong, and it is more common for users to use the namenode address directly.
+
+## Compatibility
+
+No compatibility issues at this time.
+
+## Implementation
+
+First add a  `String` type with the name `ProtocolTcp` in `endpoint`.
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-pair.md]: We don't need to add so detailed code in RFC, instead, we can describe them in clear sentences.

For example:

- Add protocol `tcp`
- Implement protocol tcp formatted (`func (p Endpoint) Tcp() (address string)`)
- Implement protocol tcp parser (`func Parse(cfg string)`)

File: docs/rfcs/8-add-tcp-pair.md (hunk 2 above)
Comment (bokket) [reply docs/rfcs/8-add-tcp-pair.md]: ok

Comment (Xuanwo) [reply docs/rfcs/8-add-tcp-pair.md]:
```suggestion
- Tracking Issue: [beyondstorage/go-endpoint/issues/9](https://github.com/beyondstorage/go-endpoint/issues/9)
```

File: docs/rfcs/8-add-tcp-protocol.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+Now we don't have a pair operation on the `hdfs address` or tcp-like operation
+
+## Compatibility
+
+No compatibility issues at this time.
+
+## Implementation
+
+- Add protocol `tcp`
+- Implement protocol tcp formatted (`func (p Endpoint) Tcp() (address string)`)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]: How about using `TCP()` as a function name instead of `Tcp()`? Keep the same style with `HTTP()` and `HTTPS`.

Comment (Xuanwo) [reply docs/rfcs/8-add-tcp-protocol.md]: And returning `address string` is not a good idea to me, how about returning `(addr, host string, port int)` instead?

File: docs/rfcs/8-add-tcp-protocol.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+- Implement protocol tcp parser (`func Parse(cfg string) (Provider, error)`)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]: I think `Parse` should return `(p Endpoint, err error)` instead.

File: docs/rfcs/8-add-tcp-protocol.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+- Implement protocol tcp object (`func NewTcp(address string) Endpoint `)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]: How about changing the API to `NewTCP(host string, port int) Endpoint` instead?

**Comment by Xuanwo** [comment]
I will approve this proposal with reviewed changes.

File: docs/rfcs/8-add-tcp-protocol.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+- Author: bokket  <bokkett@gmail.com>
+- Start Date: 2021-07-17
+- RFC PR: [beyondstorage/go-endpoint#8](https://github.com/beyondstorage/go-endpoint/pull/8)
+- Tracking Issue: [beyondstorage/go-endpoint/issues/9](https://github.com/beyondstorage/go-endpoint/issues/9)
+
+# RFC-8: Add Tcp protocol
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]:
```suggestion
# RFC-8: Add TCP protocol
```

File: docs/rfcs/8-add-tcp-protocol.md (hunk 7 above)
Comment (bokket) [reply docs/rfcs/8-add-tcp-protocol.md]: ok

File: docs/rfcs/8-add-tcp-protocol.md (hunk 8 above)
Comment (bokket) [reply docs/rfcs/8-add-tcp-protocol.md]: my mistake

File: docs/rfcs/8-add-tcp-protocol.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+Now we don't have a pair operation on the `hdfs address` or tcp-like operation
+
+## Compatibility
+
+No compatibility issues at this time.
+
+## Implementation
+
+- Add protocol `tcp`
+- Implement protocol tcp formatted (`func (p Endpoint) TCP() (addr,host string,port int)`)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]:
```suggestion
- Implement protocol tcp formatted (`func (p Endpoint) TCP() (addr, host string, port int)`)
```

File: docs/rfcs/8-add-tcp-protocol.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+hdfs usually use the `New(address string)` method to access a namenode node, the user will be the user running the code. If the address is an empty string, it will try to get the NameNode address from the Hadoop configuration file.
+
+## Proposal
+
+I suggest adding a tcp protocol to allow the user to specify the address.
+
+It likes `tcp:<host>:<port>`
+
+- The `type` of `tcp` should be `String` and is a `const`
+- The `alias` of `ProtocolTCP` should follow  [go-endpoint](https://github.com/beyondstorage/go-endpoint/blob/master/README.md)
```
Comment (Xuanwo) [inline docs/rfcs/8-add-tcp-protocol.md]:
```suggestion
```

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
Thanks for your contribution!

**Pull request merged**